core/Dispatch.cpp

```diff
@@ -146,86 +193,134 @@ namespace olympia
             InstPtr & ex_inst_ptr = dispatch_queue_.access(0);
             Inst & ex_inst = *ex_inst_ptr;
-            const auto target_unit = ex_inst.getUnit();
+            const auto target_pipe = ex_inst.getPipe();
 
-            sparta_assert(target_unit != InstArchInfo::TargetUnit::UNKNOWN,
+            sparta_assert(target_pipe != InstArchInfo::TargetPipe::UNKNOWN,
                           "Have an instruction that doesn't know where to go: " << ex_inst);
 
-            if(SPARTA_EXPECT_FALSE(target_unit == InstArchInfo::TargetUnit::ROB))
+            if (SPARTA_EXPECT_FALSE(target_pipe == InstArchInfo::TargetPipe::SYS))
             {
                 ex_inst.setStatus(Inst::Status::COMPLETED);
                 // Indicate that this instruction was dispatched
                 // -- it goes right to the ROB
                 dispatched = true;
             }
-            else {
+            else
+            {
                 // Get the dispatchers used to dispatch the target.
                 // Find a ready-to-go dispatcher
-                auto & dispatchers = dispatchers_[target_unit];
-                for(auto & disp : dispatchers)
+                sparta_assert(static_cast<size_t>(target_pipe) < InstArchInfo::N_TARGET_PIPES,
+                              "Target pipe: " << target_pipe
+                                              << " not found in dispatchers, make sure pipe is "
+                                                 "defined and has an assigned dispatcher");
+                auto & dispatchers = dispatchers_[static_cast<size_t>(target_pipe)];
+                // so we have a map here that checks for which valid dispatchers for that
+                // instruction target pipe map needs to be: "int": [exe0, exe1, exe2]
+                if (target_pipe != InstArchInfo::TargetPipe::LSU)
                 {
-                    if(disp->canAccept())
+                    uint32_t max_credits = 0;
+                    olympia::Dispatcher* best_dispatcher;
+                    // find the dispatcher with the most amount of credits, i.e the issue queue with
+                    // the least amount of entries
+                    for (auto & dispatcher_iq : dispatchers)
                     {
-                        disp->acceptInst(ex_inst_ptr);
-                        ++unit_distribution_[target_unit];
-                        ++(unit_distribution_context_.context(target_unit));
-                        ++(weighted_unit_distribution_context_.context(target_unit));
+                        if (dispatcher_iq->canAccept() && dispatcher_iq->getCredits() > max_credits)
+                        {
+                            best_dispatcher = dispatcher_iq.get();
+                            max_credits = dispatcher_iq->getCredits();
+                        }
+                    }
+                    if (max_credits != 0)
+                    {
+                        best_dispatcher->acceptInst(ex_inst_ptr);
```

Comment:
Are you sure `best_dispatcher` will be valid here?

Reply:
Added to check to make sure `best_dispatcher` is not nullptr in [19ed4b7](https://github.com/riscv-software-src/riscv-perf-model/pull/140/commits/19ed4b7911a0f115a2f9e2c2c19e5ad3324bbede)